[ADD] *: practice task implementations

ijja-odoo opened the pull request:
Task: Event Ticket Limiter 

Added a configurable maximum ticket limit per registration in events, updating ticket views and enforcing the limit dynamically on the website.

Task: Product Warranty Management

Added product warranty management, allowing users to enable warranties on products, define warranty configuration details and manage them from the sales configuration menu. A wizard is provided to easily add warranties to eligible products. Access controls are included for security.

Task: Supplier Portal

Added a supplier portal for uploading bills in PDF and XML formats. Suppliers can select associated companies, and upon upload, a draft bill is generated. Basic file validation ensures only valid files are accepted.

Commit:
[ADD] event_ticket_limiter: add configuration to limit max tickets per event

Added a new configuration to set a maximum limit on the number of tickets a
single attendee can purchase per registration.

- Introduced a setting on event tickets to define the maximum tickets allowed
  per registration.
- Updated the event management interface to include this limit.
- Applied the restriction dynamically on the event registration page, replacing
  the previous fixed condition.

Task: 4598794

Commit:
[ADD] product_warranty: add product warranty with management and configuration

Introduced product warranty, allowing management and configuration of warranties
for products.

- Added an option to enable or disable warranties for products.
- Introduced a warranty configuration with details like name, percentage, and
  duration.
- Provided a dedicated menu for managing warranties under sales settings.
- Implemented a wizard to easily add warranties to eligible products.
- Defined necessary access rules for proper security control.

Task: 4602870

robodoo commented:
[![Pull request status dashboard](https://mergebot.odoo.com/odoo/tutorials/pull/418.png)](https://mergebot.odoo.com/odoo/tutorials/pull/418)

aktr-odoo commented:
Update commit message with a more appropriate one.

ijja-odoo commented:
@aktr-odoo Changed the commit message, could you please have a look?

Commit:
[ADD] supplier_portal: portal for suppliers to create bills from website

- Added a supplier portal on the website, allowing suppliers to log in and
  upload bills in PDF and XML formats.
- Added company selection in the portal, displaying only companies the supplier
  is associated with.
- Upon successful file upload, a new bill is generated in the draft state,
  linked to the selected company and supplier.
- Implemented basic file validation to ensure only valid XML and PDF files are
  accepted.

Task: 4619119

aktr-odoo commented on product_warranty/views/product_views.xml:
give proper file name

aktr-odoo commented on product_warranty/views/sale_menus.xml:
Give proper file name

aktr-odoo commented on product_warranty/tests/test_add_warranty.py:
All the best cases are included in one file.
The test cases must be in a separate file, code-wise.

aktr-odoo commented on product_warranty/tests/test_add_warranty.py:
```diff
@@ -0,0 +1,72 @@
+            'sale_order_line_id': self.so_line_with_warranty.id
+        })
+        
+        exptected_order_lines = len(self.sale_order.order_line)+1
+        wizard.action_add_warranty()
+        self.assertEqual(len(self.sale_order.order_line), exptected_order_lines, "Warranty line must be added to the sale order")
```
Check the length doesn't seem appropriate to me.
What do you think?

ijja-odoo replied:
What I thought was as tests have their own flow of running checking length will work as no other operations will be going on. But the test can be improved by checking the order lines instead of just checking the length.

aktr-odoo commented on product_warranty/tests/test_add_warranty.py:
```diff
@@ -0,0 +1,72 @@
+            'sale_order_line_id': self.so_line_with_warranty.id
+        })
+        
+        expected_order_lines = len(self.sale_order.order_line)
+        wizard.action_add_warranty()
+        self.assertEqual(len(self.sale_order.order_line), expected_order_lines, "Warranty line with no warranty configuration must not be added to sale order")
```
Same here
Check the length doesn't seem appropriate to me.

aktr-odoo commented on product_warranty/wizard/add_warranty.py:
Give proper file name

aktr-odoo commented on product_warranty/wizard/add_warranty_views.xml:
Give proper file name

aktr-odoo commented on product_warranty/wizard/warranty_line.py:
Give proper file name

aktr-odoo commented on product_warranty/wizard/add_warranty.py:
```diff
@@ -0,0 +1,44 @@
+
+
+class AddWarranty(models.TransientModel):
+    _name = "add.warranty"
```
Give proper name.

aktr-odoo commented on product_warranty/wizard/add_warranty.py:
```diff
@@ -0,0 +1,44 @@
+            res["warranty_line_ids"] = [
+                Command.create({"sale_order_line_id": order_line.id})
+                for order_line in order_lines
+                if order_line.product_template_id.is_warranty_available
```
Can this be done without using `if`?

ijja-odoo replied:
Yes, an alternate approach without using `if` will be using `filtered()` function.

aktr-odoo commented on product_warranty/wizard/add_warranty.py:
```diff
@@ -0,0 +1,44 @@
+                }))
+
+        if order_lines:
+            self.sale_order_id.write({"order_line": order_lines})
```
Its better to create order lines rather than appending and then writing

aktr-odoo reviewed (commented):
:wave: 

I have left a few comments. Could you please have a look?

ijja-odoo commented:
@aktr-odoo Requested changes have been made can you please have a look?

aktr-odoo commented on supplier_portal/controllers/main.py:
```diff
@@ -0,0 +1,62 @@
+        bill_xml = vals.get("bill-type-xml")
+        bill_pdf = vals.get("bill-type-pdf")
+        
+        if company_id not in request.env.user.company_ids.mapped("id"):
+            return self._acknowledge_supplier("danger", "Error", "Invalid company ID provided")
```
The template only allows to select user company. So do we need to check this?

ijja-odoo replied:
Request payload can be modified, so isn't checking this helps preventing unauthorized bill uploads?

aktr-odoo commented on supplier_portal/controllers/main.py:
```diff
@@ -0,0 +1,62 @@
+            "partner_id": request.env.user.partner_id.id,
+        })
+
+        if bill_xml:
+            self._attach_file(account_move.id, bill_xml.filename, bill_xml.read(), "text/xml")
+            
+        if bill_pdf:
+            self._attach_file(account_move.id, bill_pdf.filename, bill_pdf.read(), "application/pdf")
```
Logic can be improved here.
`Create` will be triggered twice here

aktr-odoo commented on supplier_portal/controllers/main.py:
```diff
@@ -0,0 +1,62 @@
+        })    
+
+    def _is_valid_file(self, file, expected_type, expected_file_extension):
+        return "." in file.filename and file.filename.split(".")[-1] == expected_file_extension and file.content_type == expected_type
```
Bad logic.
HINT: This can be done with external library

aktr-odoo reviewed (changes requested):
:wave: 

I have left a few comments. Could you please have a look?

ijja-odoo commented:
@aktr-odoo Requested changes have been made can you please have a look?

Commit:
[ADD] l10n_in_custom_duty: configuration of custom import duty on bills

PURPOSE

This commit introduces the Indian localization module `l10n_in_custom_duty`,
allowing businesses to configure and manage custom import duties on vendor bills
within the Accounting module.

SPECIFICATION

- Added a Custom Import & Export Duty configuration under the Indian Integration
  section in Accounting settings.
- Introduced a Bill of Entry button in vendor bills, which opens a wizard to add
  custom duty, calculated based on the custom currency rate.
- On confirming the wizard, a new journal entry is created and linked to the
  vendor bill.

khva-odoo commented on l10n_in_custom_duty/wizard/add_custom_duty.py:
```diff
@@ -0,0 +1,190 @@
+        return res
+
+    active_bill_id = fields.Many2one(comodel_name="account.move")
+    ref_bill_id = fields.Many2one(comodel_name="account.move", string="Reference", readonly=True)
```
can improve the name of this field, to be used it for more generic purpose after removing `active_bill_id`

khva-odoo commented on l10n_in_custom_duty/wizard/add_custom_duty.py:
```diff
@@ -0,0 +1,190 @@
+        """
+        self.ensure_one()
+        
+        custom_duty_account_id = self.active_bill_id.company_id.import_custom_duty_account_id
+        default_tax_account_id = self.active_bill_id.company_id.import_default_tax_account_id
```
```suggestion
        custom_duty_account_id = self.ref_bill_id.company_id.import_custom_duty_account_id
        default_tax_account_id = self.ref_bill_id.company_id.import_default_tax_account_id
```

khva-odoo commented on l10n_in_custom_duty/wizard/add_custom_duty.py:
```diff
@@ -0,0 +1,190 @@
+            }),
+        ]
+        
+        for custom_duty_line in self.add_custom_duty_line_ids:
+            taxes = custom_duty_line.tax_ids.compute_all(custom_duty_line.taxable_amount, currency=self.currency_id)
+            
+            for tax_res in taxes["taxes"]:
+                tax = self.env["account.tax"].browse(tax_res["id"])
+                debit_account = tax.invoice_repartition_line_ids.filtered(lambda line: line.repartition_type == "tax" and line.account_id).account_id
+                
+                if not debit_account:
+                    raise UserError(f"No debit account found for tax {tax.name}")
```
this logic can be simplified.

khva-odoo commented on l10n_in_custom_duty/wizard/add_custom_duty.py:
File name can be improved.

khva-odoo commented on l10n_in_custom_duty/wizard/add_custom_duty_line.py:
File name can be improved.

khva-odoo commented on l10n_in_custom_duty/wizard/add_custom_duty_line.py:
```diff
@@ -0,0 +1,42 @@
+class AddCustomDutyLine(models.TransientModel):
+    _name = "add.custom.duty.line"
+    _description = "Add Custom Duty Line"
+    _transient_max_hours = 0
```
any specific reason for using this?

ijja-odoo replied:
Since the model is `TransientModel`, records are removed by `ir.autovacuum`. To retain details for `Bill of Entry`, I added this but will try another approach.

khva-odoo commented on l10n_in_custom_duty/wizard/add_custom_duty_line.py:
```diff
@@ -0,0 +1,42 @@
+from odoo import api, fields, models
+
+
+class AddCustomDutyLine(models.TransientModel):
+    _name = "add.custom.duty.line"
```
Naming conventions can be improved.

khva-odoo commented on l10n_in_custom_duty/wizard/add_custom_duty.py:
```diff
@@ -0,0 +1,190 @@
+from odoo.exceptions import UserError, ValidationError
+
+
+class AddCustomDuty(models.TransientModel):
+    _name = "add.custom.duty"
```
Naming conventions can be improved.

khva-odoo commented on l10n_in_custom_duty/wizard/add_custom_duty.py:
```diff
@@ -0,0 +1,190 @@
+            self.bill_of_entry_id.write({
+                "line_ids": [Command.clear()] + line_vals,
+            })
+            bill_of_entry_journal = self.bill_of_entry_id
```
misleading variable name.

khva-odoo commented on l10n_in_custom_duty/wizard/add_custom_duty_line.py:
```diff
@@ -0,0 +1,42 @@
+    ]
+    
+    account_move_line_id = fields.Many2one(comodel_name="account.move.line")
+    add_custom_duty_id = fields.Many2one(comodel_name="add.custom.duty")
```
can improve the naming convention.

khva-odoo reviewed (commented):
Hey,
Can you please have a look into the comments??

ijja-odoo commented:
@khva-odoo Requested changes have been made, can you please have a look?
